echo-pubsub/src/main/java/com/netflix/spinnaker/echo/pubsub/google/GooglePubsubSubscriber.java

```diff
@@ -0,0 +1,153 @@
+
+  private String topic;
+
+  @Getter
+  private Subscriber subscriber;
+
+  private Integer ackDeadlineSeconds;
+
+  static private final PubsubType pubsubType = PubsubType.GOOGLE;
+
+
+  public GooglePubsubSubscriber(String name, String project, Subscriber subscriber, Integer ackDeadlineSeconds) {
+    this.name = name;
+    this.project = project;
+    this.topic = String.format("projects/%s/topics/%s", project, name);
+    this.subscriber = subscriber;
+    this.ackDeadlineSeconds = ackDeadlineSeconds;
+  }
+
+  @Override
+  public PubsubType pubsubType() {
+    return pubsubType;
+  }
+
+  @Override
+  public String topic() {
+    return topic;
+  }
+
+  public static GooglePubsubSubscriber buildSubscriber(String name,
+                                                       String project,
+                                                       String jsonPath,
+                                                       Integer ackDeadlineSeconds,
+                                                       PubsubMessageCache pubsubMessageCache) {
+    Subscriber subscriber;
+    GooglePubsubMessageReceiver messageReceiver = new GooglePubsubMessageReceiver(ackDeadlineSeconds, name, pubsubMessageCache);
+
+    if (jsonPath != null && !jsonPath.isEmpty()) {
+      Credentials credentials = null;
+      try {
+        credentials = ServiceAccountCredentials.fromStream(new FileInputStream(jsonPath));
+      } catch (IOException e) {
+        log.error("Could not import Google pubsub json credentials: {}", e.getMessage());
+      }
+      subscriber = Subscriber
+          .defaultBuilder(SubscriptionName.create(project, name), messageReceiver)
+          .setCredentialsProvider(FixedCredentialsProvider.create(credentials))
+          .setMaxAckExtensionPeriod(Duration.ofSeconds(0))
+          .build();
+    } else {
+      subscriber = Subscriber.defaultBuilder(SubscriptionName.create(project, name), messageReceiver).build();
+    }
+
+    subscriber.addListener(new GooglePubsubFailureHandler(), MoreExecutors.directExecutor());
+
+    return new GooglePubsubSubscriber(name, project, subscriber, ackDeadlineSeconds);
+  }
+
+  private static class GooglePubsubMessageReceiver implements MessageReceiver {
+
+    private Integer ackDeadlineSeconds;
+
+    private PubsubMessageCache pubsubMessageCache;
+
+    private String subscriptionName;
+
+    private NodeIdentity identity = new NodeIdentity();
+
+    public GooglePubsubMessageReceiver(Integer ackDeadlineSeconds,
+                                       String subscriptionName,
+                                       PubsubMessageCache pubsubMessageCache) {
+      this.ackDeadlineSeconds = ackDeadlineSeconds;
+      this.subscriptionName = subscriptionName;
+      this.pubsubMessageCache = pubsubMessageCache;
+    }
+
+    @Override
+    public void receiveMessage(PubsubMessage message, AckReplyConsumer consumer) {
+      String messagePayload = message.getData().toStringUtf8();
```

Comment:
I wonder if we can generify some of this, since the other pubsub-style subscribers can benefit from this careful deduping logic.